Add fractribution code

zloff opened the pull request:
<!--
If this is your first time contributing you will be asked to sign the Individual Contributor License Agreement.
If you would prefer to read this in advance of submitting your PR you can find it here https://docs.google.com/forms/d/e/1FAIpQLSd89YTDQ1XpTZbj3LpOkquV_h1Y8k9ay3iFbJsZsJrz18I23Q/viewform
-->

## Description & motivation
<!--
Add Fractribution code to use with the marketing attribution accelerator https://github.com/snowplow/marketing-attribution-accelerator/pull/2 

I will add some sample data that can be used with this to the marketing attribution pr early next week.

I have added the same reviewers as the Marketing Attribution Accelerator, but please add (or remove!) reviewers as appropriate. 
-->

## Checklist
- [ x] I have verified that these changes work locally
- [x ] I have updated the README.md (if applicable)
- [ ] I have added tests & descriptions to my models (and macros if applicable)
- [ ] I have raised a [documentation](https://github.com/snowplow/documentation) PR if applicable (Link here if required)

<!-- 
## Release Only Checklist
- [ ] I have updated the version number in all relevant places
- [ ] I have update the Github pages `dbt docs` 
- [ ] I have updated the CHANGELOG.md 
-->

zloff committed:
Add fractribution code

agnes-kiss commented on dbt_project.yml:
```diff
@@ -0,0 +1,50 @@
+
+name: 'fractribution'
```
I know you amended everything to be aligned to `fractribution` in the other PR so sorry about this, but now that it will live inside the official snowplow dbt package realm can it be renamed to `snowplow_fractribution` to align it to our naming convention? (together with the other places like at the adapter dispatch bits etc.)

zloff replied:
No worries!

agnes-kiss commented on dbt_project.yml:
```diff
@@ -0,0 +1,50 @@
+config-version: 2
+
+# This setting configures which "profile" dbt uses for this project.
+profile: 'fractribution'
```
Can this be `default` instead? Again, just for the standard approach.

agnes-kiss commented on dbt_project.yml:
```diff
@@ -0,0 +1,50 @@
+on-run-start: "{{ create_udfs() }}"
+models:
+  fractribution:
+    +schema: "fractribution"
```
I'm still thinking we should perhaps put everything against default `derived` schema (as you said there is no point separating some of them to scratch) just so that if our customers use multiple snowplow packages they would probably expect  them to be all in the same schemas (like our other packages), especially as the source tables would also be found inside this schema. The models would all need to start with `snowplow_fractribution_` for it to work but it would be clearly separated that way (and users won't have to change profiles either). What do you think?

zloff replied:
Sounds good - i'll update everything here and in the accelerator.

agnes-kiss commented on models/snowflake/conversions_by_customer_id.sql:
```diff
@@ -0,0 +1,24 @@
+    END AS customerId,
+    derived_tstamp AS conversionTimestamp,
+    {{ conversion_value() }} AS revenue
+FROM
```
I would say this model should be referring to an incremental model and not the atomic.events table directly and as such would be best to place inside the dbt project the user handles when operating the snowplow_web package. Otherwise it would be an expensive operation, especially for larger lookback-windows. It would utilise the `this-run` tables to add incrementally all the events that generated revenue in a performant way. That model could be added to this package as a sample model that you would encourage users to apply and do a backfill incrementally first, before running this package. There is a way to only rebuild this using the base module and disabling the rest of the web package to save costs. A similar concept happens in the snowplow_media_player package but can explain separately, if needed.

Optionally you could still provide a sample drop and recompute model you added here so the choice would be on the user to use whichever way they want, although you would probably need to add a new variable and decide the logic based on that (as the source would be different that needs to be handled in that case).

zloff replied:
Thanks Agnes. I think I understand that you would not recommend querying atomic.events, but I'm a bit confused by the rest. In terms of expense, i'm not sure that it would be overly expensive -as only a few columns are being selected, even if the date range is large. 
Happy to do what you recommend, but just a little confused as to what that is exactly!

zloff replied:
Hi @agnessnowplow, if you could elaborate on this as discussed on Friday, maybe with some examples or something to help us understand what you are suggesting? That would be very helpful :) Thanks

agnes-kiss replied:
For sure. So I would make it possible for the user to either use the `atomic.events` as a source for conversions as you suggested or they would put their own choice of model as a source within the FROM clause through a variable. The recommended suggestion to use for this would be one of the derived tables of the new ecommerce model that is currently being finalised, most likely this one, which contains the revenue generating transaction events by users: 
<img width="367" alt="image" src="https://user-images.githubusercontent.com/95634439/204624080-f4483c90-6141-4631-801c-33a225221833.png">
If they do that then there is no need to query the event table again just for these transactions, those events will already be available in a much smaller source table. They could also create their own custom incremental model to use as a source, if they wanted to, as I highlighted earlier - as a third option. (Of course they would have to make sure that the data they have is already up-to-date with the range they are querying.) How it could work, for instance, is there would be a new variable called `conversion_source` with atomic and events from the source by default ( `conversion_source: "{{ source('atomic', 'events') }}"`) We would then instruct the user to overwrite this source variable in their own dbt_project.yml if they want to and use their preferred model instead. Let me know if this helps / your thoughts. Thanks!

zloff replied:
Ahhh I understand now, thanks so much for taking the time to write all this out. That's a good idea, I'll make the necessary changes :)

zloff replied:
I've made the changes to this and also in the modelling_3.md file in the other repo, hopefully these changes are what you were wanting!

agnes-kiss replied:
Nice one! You opted for a slightly different solution as I thought of theoretically speaking but I like it better as you reduced the number of variables!

agnes-kiss commented on dbt_project.yml:
```diff
@@ -0,0 +1,50 @@
+
+vars:
+  conversion_window_start_date: '2022-06-03' # conversion window start date
+  conversion_window_end_date: '2022-08-01' # conversion window end date
```
Was wondering if it is worth the hassle to define the current date as the default end date. It could be called 'current_date' and the logic could be if the `conversion_window_end_date` = 'current_date' (or perhaps leave it as '' and refer to that) it would retrieve the current_date() (as it does not work within the project.yml itself), otherwise just take the user input as is. It all depends on how likely this scenario is, really.

zloff replied:
Thanks Agnes, this is a good point. Mike suggested current_date()-1 may be better since current_date won't have all the data yet. I will make the changes for this.

agnes-kiss replied:
Great, that seems reasonable. Also good idea on the -31 days for the start date!

agnes-kiss commented on models/snowflake/channel_counts.sql:
```diff
@@ -0,0 +1,10 @@
+    campaign,
+    source,
+    medium,
+    COUNT(*) AS number_of_events
```
As much as I understand we are talking about the number of page_view events in each cases, right? Would that make sense to rename this to `number_of_pageviews`? Just so it is easier to deduct what is being counted without knowing how it works altogether.

zloff replied:
Although we are using page_view events, we aren't really thinking of them as page views, as they are just events with associated marketing parameters, either multiple events or one (depending on the value of consider_intrasession_channels). So this would more accurately be number_of_sessions, as each event is being treated as a session, in the GA sense of a session. (Also see comment below)

agnes-kiss commented on models/snowflake/sessions_by_customer_id.sql:
```diff
@@ -0,0 +1,43 @@
+SELECT
```
The name of this model confused me before I read the readme and understood that it is google's naming that is being followed here if I'm not mistaken. I wonder if it would make sense to rename this to `page_views_by_customer_id`?

zloff replied:
This is sort of related to my above comment, and yes it is to follow Google's conventions. 

It's confusing, there are kind've two notions of a session - the GA type of session where if you change your marketing attribution you start a new session automatically, and the snowplow session, where a new channel doesn't mean a new session is created (unless 30min of inactivity). 

So in this model, we are returning one page view per session if they are not considering intrasession channels. We are returning more than one page view per session if they are considering intrasession channels - in this case, each event  ‘page view’ is considered to be a separate session and is treated as such. So, the rows returned by this model are associated with page views, but we don’t treat them as page_views, we treat them as separate sessions, in the GA sense. Mike kept consistency of naming to make it more of a reproduction of the fractribution stuff. 

Let me know if this is still confusing, Mike also explained it to me (he wrote the bulk of this code) so I may not have explained it well enough!

agnes-kiss replied:
Thank you very much for the detailed explanation, it is much clearer! One thing that I wanted to mention earlier but forgot is that we do not keep the detailed documentation inside the models anymore but shifted anything that are not the absolute essentials to https://github.com/snowplow/documentation. We are currently changing the structure slightly within the data modelling sections, but what I can do is once that is finalised in a day or two I will open a new draft PR, create a placeholder folder with the related index.md file and copy the currently existing readme contents there. Once there, perhaps it would make sense to explain anything special that the users will need to know about understanding the model and how to operate it (e.g. custom, more advanced used cases). Would you be happy to add such extra documentation and / or review the PR together?

zloff replied:
Yes of course, thanks for your help!

agnes-kiss commented on models/snowflake/sessions_by_customer_id.sql:
```diff
@@ -0,0 +1,43 @@
+
+    {% if var('consider_intrasession_channels') %}
+        -- yields one row per channel change
+        AND mkt_medium IS NOT NULL AND mkt_medium != ''
```
It might be I am missing a major point, so please correct me I'm wrong but I feel like we are trying to filter out the last-touch logic earlier on if relevant and at a different stage ask the user to decide on the attribution model (again) via the python code? Would it work if instead of an if-else statement we just included the page_view_in_session ratio in this model and do the filtering later on without the variable?

zloff replied:
We're not actually filtering out the last touch, we're trying to figure out ‘what logic would the user like to use to attribute a session to its marketing parameters?’, which is separate from the fractribution stuff - that says, give me your session and its attribution. This is saying, for a given session, how would you like to attribute that session? Would you like to attribute that session to the first marketing parameters? This isn’t last or first touch, the only scope of the function is to go - this is the session, how do you want to attribute it (look at the first event?) OR if the marketing parameters change within the session, then go and use those instead and have multiple sessions per snowplow session. 
Once again, let me know if you'd like further explanation :)

agnes-kiss commented on models/snowflake/sessions_by_customer_id.sql:
```diff
@@ -0,0 +1,43 @@
+    )
+
+    {% if var('consider_intrasession_channels') %}
+        -- yields one row per channel change
```
This comment confused me a bit, because of the `derived_tstamp` each page_view will be presented, not just those with medium "change" but I guess we want to capture each page_views where there is a mkt_medium defined, right?

zloff replied:
Yep!

agnes-kiss commented on models/snowflake/channel_spend.sql:
```diff
@@ -0,0 +1,14 @@
+
+
+-- TODO: put in your own spend calculations per channel in the channel_spend macro in your own dbt project
+-- the model assigns an example 10k spend to each channel found in channel_counts
```
Perhaps it is worth to include a sample query what it may actually look like for the user. When I first had a look I did not understand why we need to collect the array first and flatten it out. This way you can suggest to them that it is a matter of a simple `select distinct` and later `case when` statements that they can completely overwrite in their macro anyway. Not sure if this needs to be in a macro or they would just have to create this model on their own project but perhaps it is better to leave it as a macro. (at least it could work out of the box for illustrational purposes even if they do not overwrite it at first)

zloff replied:
I added a very small sample query in the channel_spend macro (just to show that you don't _have_ to do any funky ARRAY_AGG, which could be confusing) and also an example output as Mike suggested on Friday, let me know if you think it's too simple though and I can change it.

agnes-kiss replied:
I think it is very obvious for the user what and how they would need to go about this, I'm sure many of them would not need it but for the few that it helps it is worth the effort! I also like the example output!

agnes-kiss commented on models/snowflake/paths_to_conversion.sql:
```diff
@@ -0,0 +1,27 @@
+
+SELECT
+  conversions_by_customer_id.customerId,
```
NIT: can this be `c` or `con` instead of `conversions_by_customer_id` (same as with the other table)? Just find it harder to read if it is this long but it might just be personal preference so not strong on this one.

agnes-kiss commented on macros/channel_classification.sql:
```diff
@@ -0,0 +1,69 @@
+        WHEN LOWER(mkt_medium) = 'referral' THEN 'Referral'
+        WHEN LOWER(mkt_medium) = 'email' THEN 'Email'
+        WHEN
+            LOWER(mkt_medium) IN ('cpc', 'ppc', 'cpv', 'cpa', 'affiliates')
```
The first two in the list - cpc, ppc - are superfluous, I think they will never get evaluated as they will be used above as 'paid_search_other' if not further up in the logic.

zloff replied:
I think Mike mentioned on Friday that this was to keep parity with Google's code

agnes-kiss reviewed (commented):
Thanks for sharing the sample events, Zoe, it has been really helping me getting a much deeper understanding for the review! I have included a couple of suggestions below, partially on how to make it more Snowplow-y but overall it is looking good and everything worked as expected!

zloff committed:
Change name to snowplow_fractribution, change schema to derived, add snowplow_fractribution_ to model names

zloff committed:
Change set dates to use current_date() as default, rename model files

zloff committed:
Add example table to model comments, rename column in channel_counts

zloff committed:
Update channel_spend information

zloff committed:
Simplify process for user to query a table other than atomic events for conversion values

agnes-kiss commented on models/snowflake/schema.yml:
```diff
@@ -3,15 +3,13 @@ version: 2
-        identifier: "{{ var('events_table', 'events') }}"
   - name: derived
-    schema: "{{ var('page_views_schema', 'derived') }}"
+    schema: atomic_derived
```
Could this be `derived` instead? It will much more likely to be in their derived schema.

zloff replied:
yes sure!

agnes-kiss reviewed (commented):
Thanks for making the changes, @zloff! Just a tiny recommended change and I think we are pretty much there! I have opened a [draft PR ](https://github.com/snowplow/documentation/pull/209)for the docs site and already copied some of the documentations either from the README or from the Accelerator. I think the dedicated overview under the Fractrubution model, Quickstart, Operation and Configuration pages need to be covered mainly. I will update them as soon as I have time but feel free to add any content you'd like or think is needed.
 
<img width="291" alt="image" src="https://user-images.githubusercontent.com/95634439/205057755-45968742-f654-4479-a18c-fe00d380a97e.png">

zloff committed:
Change schema to derived

agnes-kiss commented on dbt_project.yml:
```diff
@@ -0,0 +1,47 @@
+
+name: 'snowplow_fractribution'
+version: '1.0.0'
```
For the first release we normally go for a minor version and increment them with each feature (let's say when we add a new adapter support). I'd suggest to start with v 0.1.0. We also have a [guide](https://github.com/snowplow-incubator/engineering-resources/blob/master/release-process.md) on the release process that we try to keep, although some things we are more relaxed about, I will try to write down each step that are relevant. 

Just so we follow the usual process, could you please create a release branch (release/snowplow-fractribution/0.1.0) and change this pr request to be merged with that first, instead of main, please? Once it is approved, you can then squash the commits together into one commit during the merge to the release branch (if not allowed for some reason you can perhaps rebase them into one commit and force push to this branch or do the same on the release branch, too) and for the commit message just say something along the lines of `Add fractribution model for Snowflake.` After this one commit is merged to the release branch, a `Prepare for release` commit is added with an entry in the CHANGELOG to specify the details. See an example of how this works [here](https://github.com/snowplow/dbt-snowplow-web/pull/3/commits). There are a couple of more steps, it is perhaps easier to just copy paste a step by step guide:

1. Create a commit on the `release/snowplow-X/Y.Z` branch with message `Prepare for release` with an update to:
    1. The `CHANGELOG` providing the information that should roughly match the commit messages into the branch (including issues closed). Add `(Thanks to @gh_contributor)`  if the commit comes from an external contributor 
    2. The `README.md` to update the version in the adaptor support
    3. The `dbt_project.yml` to update the version
    4. Any custom example or integration test project yaml files (usually version number)
    5. Update the docs to ensure they match the latest version of the branch (run on redshift where possible) 
        
        ```bash
        dbt docs generate
        rm -f docs/catalog.json
        rm -f docs/manifest.json
        rm -f docs/run_results.json
        cp target/catalog.json docs/catalog.json
        cp target/manifest.json docs/manifest.json
        cp target/run_results.json docs/run_results.json
        ``` 
2. Create a merge request from the release branch to the main/master branch
3. Once reviewed, merge into the main/master branch
4. Checkout the branch locally and tag it with the release number:
    1. `git checkout main`
    2.  `git pull`
    3. `get tag 0.x.y`
    4. `git push --tags`
5. Create a Github Release from this tag with:
    1. **Summary**: A brief overview of the release
    2. **Features**: A list of features, should roughyly match the commit messages
    3. **Upgrading**: Information about how to upgrade to the latest version and/or take advantage of new features
6. Merge the relevant PR in the [documentation site](https://github.com/snowplow/documentation) if not already done
7. Delete all branches created as part of the release
8. If this is the first release of a new dbt package, it also needs to be added to the dbt hub. Raise a Pull Request in [https://github.com/dbt-labs/hubcap/ ](https://github.com/dbt-labs/hubcap/)with changes to the hub.json file.

Sorry for this long post but was not sure how different your processes are to this so best to keep it verbose. Please let me know if you need more details. Thanks!

zloff replied:
Thanks Agnes! I haven't really done this before, so the long post is very helpful! I have made the branch and pointed this PR to that branch. From what I understand, I perform the rest of these steps after this PR has been merged to that branch?

agnes-kiss replied:
Yes, exactly! This is an example PR for the dat-labs repo as well, if it helps: https://github.com/dbt-labs/hubcap/pull/115/commits/502bd140fa4de78453c238434f22c052b7d138b0.

agnes-kiss commented on README.md:
```diff
@@ -9,6 +9,33 @@
+1. Ensure the setup steps have been completed above.
+2. Run `dbt run`, or `dbt run --select package:fractribution`
+
+
```
Can we have an Adapter support section as well here, please? It is important for users to know it is for Snowflake and for which version etc. See example, [here](https://github.com/snowplow/dbt-snowplow-web/blob/main/README.md).

zloff replied:
Added, and also updated the Readme to make it more like the snowplow-web one. Let me know if I should change anything

zloff committed:
Update readme, change version

agnes-kiss commented on macros/create_udfs.sql:
```diff
@@ -0,0 +1,159 @@
+
+
+  -- Functions for applying transformations to path arrays.
+  -- unique: Identity transform.
```
in the end it is called uniquePath, right?

zloff replied:
Ah you're right, i'll change it to unique. Thanks!

zloff replied:
Actually, scratch that, I think Mike must have named it uniquePath because dbt run errors if it's just called Unique. I'll change the docs.

agnes-kiss replied:
Yeah, better to name it more unique, haha.

agnes-kiss commented on macros/create_udfs.sql:
```diff
@@ -0,0 +1,159 @@
+  {% set unique %}
+  -- Returns the unique/identity transform of the given path array.
+  -- E.g. [D, A, B, B, C, D, C, C] --> [D, A, B, B, C, D, C, C].
+  CREATE FUNCTION IF NOT EXISTS {{schema}}.UniquePath(path ARRAY) -- TODO: should this be called unique?
```
I think this comment is temporary and internal, it needs to be removed.

zloff replied:
Thank you, done

agnes-kiss commented on macros/create_udfs.sql:
```diff
@@ -0,0 +1,159 @@
+  --   E.g. [D, A, B, B, C, D, C, C] --> [D, A, B, C].
+  -- frequency: Removes repeat events but tracks them with a count.
+  --   E.g. [D, A, B, B, C, D, C, C] --> [D(2), A(1), B(2), C(3)).
+
```
Raised it already in the other PR, just noting that 'recency' might be missing, too.

zloff replied:
Thanks - I've removed that as it shouldn't have been there

zloff committed:
Change unique to uniquePath

agnes-kiss reviewed (commented):
Thank you for making all the changes @zloff! I think the only thing that's missing (as we were discussing in the docs site of the package) is to add the python scripts with perhaps a couple of words about it in the Readme once the questions/issues are resolved about them that have just been raised in the accelerator repo. Otherwise all good I think!

zloff committed:
Update readme and add scripts to utils folder

agnes-kiss commented on utils/main_snowplow_snowflake.py:
```diff
@@ -0,0 +1,233 @@
+# coding=utf-8
+# Copyright 2022 Google LLC..
```
```suggestion
# Copyright 2022 Snowplow, Google LLC..
```

agnes-kiss commented on utils/fractribution.py:
```diff
@@ -0,0 +1,350 @@
+# coding=utf-8
+# Copyright 2022 Google LLC..
```
```suggestion
# Copyright 2022 Snowplow, Google LLC..
```

agnes-kiss replied:
Not sure if this one was modified as well, if yes, it is better to add Snowplow as well to the copyright section.

paulboocock replied:
I'd suggest we make this `Snowplow, Google LLC` rather than `-`.

agnes-kiss commented:
One thing Ryan mentioned while reviewing the Documentation PR is that we are missing the `snowplow_dbt` query tag that we always include for any of our Snowflake-run models for visibility. Can this config section be added to the top of each model, please? For this to work we would need to add another package dependency, the `snowplow_utils`. Even though version 0.13.0 will be out very soon it is best to add this with 0.12.0 dependency as from 0.13.0 onward we will rely on dbt 1.3.0 as a minimum.
```
{{ 
  config(
    sql_header=snowplow_utils.set_query_tag(var('snowplow__query_tag', 'snowplow_dbt'))
  ) 
}}
```
```
packages:
  - package: snowplow/snowplow_utils
    version: [">=0.12.0", "<0.13.0"]
```

paulboocock commented:
I have created a docker repository for us to dockerise the scripts, it's called `snowplow/fractribution` and I've also added the secrets to the Github repo for publishing to Docker - `DOCKER_USERNAME` and `DOCKER_PASSWORD`.

I don't have a great example of another repository which publishes to Docker that feels similar to this one. But you could take a look here, which is a `Go` based tracker and has the commands visible on what we do to publish. You'd have to work out how to create the docker image appropriately in Github Actions first though (which I imagine is a case of using a base image with Python already installed and then mounting the required files into the container on build).

https://github.com/snowplow/snowplow-tracking-cli/blob/master/Makefile#L105-L109

agnes-kiss commented on README.md:
```diff
@@ -5,9 +5,84 @@
+2. Run `dbt run`, or `dbt run --select package:fractribution`
+3. Run the python script, e.g. `python utils/main_snowplow_snowflake.py --conversion_window_start_date '2022-06-03' --conversion_window_end_date '2022-08-01'` (see [doc site](https://docs.snowplow.io/docs/modeling-your-data/modeling-your-data-with-dbt/) for further details)
+
+### Differences to Fractribution
```
```suggestion
### Differences to Google's Fractribution
```

agnes-kiss commented on README.md:
```diff
@@ -5,9 +5,84 @@
+
+### Differences to Fractribution
+
+There are some changes from the [original](https://github.com/google/fractribution) Fractribution code that have been noted below.
```
```suggestion
There are some changes from [Google's](https://github.com/google/fractribution) Fractribution code that have been noted below.
```

agnes-kiss replied:
As per recommendation on the Docs' PR.

zloff committed:
Apply copyright suggestions from code review

Co-authored-by: Agnes Kiss <95634439+agnessnowplow@users.noreply.github.com>

zloff committed:
Add query tag, snowplow_utils package, update readme

zloff committed:
Update to match documentation change

agnes-kiss commented on README.md:
```diff
@@ -5,9 +5,84 @@
+
+Please refer to the [doc site](https://docs.snowplow.io/docs/modeling-your-data/modeling-your-data-with-dbt/) for a full breakdown of the package. To run this package with detailed instructions and example data, see the [fractribution accelerator](https://docs.snowplow.io/accelerators/snowplow_fractribution/).
+
+### Adaptor Support
```
```suggestion
### Adapter Support
```

agnes-kiss commented on README.md:
```diff
@@ -5,9 +5,84 @@
+
+### Requirements
+
+- A dataset of web events from the [Snowplow Javascript Tracker](https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/) and familiarity with the [snowplow-web](https://hub.getdbt.com/snowplow/snowplow_web/latest/) dbt package
```
```suggestion
- A dataset of web events from the [Snowplow Javascript Tracker](https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/) and familiarity with the [snowplow-web](https://hub.getdbt.com/snowplow/snowplow_web/latest/) dbt package
- The `snowplow_web_page_views` table generated by the snowplow-web package available in your warehouse
- A `conversions` dataset available in your warehouse (a dataset containing revenue data by users).
- A `channel spend` dataset available in your warehouse (a dataset containing total spend by channel)
```

agnes-kiss reviewed (approved):
A couple of small README change recommendations but it is ready to be merged from my side.

agnes-kiss commented on README.md:
```diff
@@ -5,9 +5,84 @@
+1. Configure the `conversion_clause` macro to filter your raw Snowplow events to successful conversion events.
+2. Configure the `conversion_value` macro to return the value of the conversion event.
+3. Configure the default `channel_classification` macro to yield your expected channels. The ROAS calculations / attribution calculations will run against these channel definitions.
+4. Configure environment variables to be used by the `connection_parameters` in the `main_snowplow_snowflake.py` file to connect to Snowflake.
```
```suggestion
4. Configure environment variables to be used by the `connection_parameters` in the `main_snowplow_snowflake.py` file to connect to Snowflake.
5. Overwrite default variables provided by the package in your dbt_project.yml file, if necessary. E.g.: make sure your `page_views_source` and `conversions_source` are aligned to what is available in your warehouse.
```

zloff committed:
Apply suggestions from code review

Co-authored-by: Agnes Kiss <95634439+agnessnowplow@users.noreply.github.com>

zloff committed:
Add dockerfile and workflow to push to Docker Hub

zloff commented:
> I have created a docker repository for us to dockerise the scripts, it's called `snowplow/fractribution` and I've also added the secrets to the Github repo for publishing to Docker - `DOCKER_USERNAME` and `DOCKER_PASSWORD`.
> 
> I don't have a great example of another repository which publishes to Docker that feels similar to this one. But you could take a look here, which is a `Go` based tracker and has the commands visible on what we do to publish. You'd have to work out how to create the docker image appropriately in Github Actions first though (which I imagine is a case of using a base image with Python already installed and then mounting the required files into the container on build).
> 
> https://github.com/snowplow/snowplow-tracking-cli/blob/master/Makefile#L105-L109

Thanks @paulboocock, @agnessnowplow - I'm assuming because you commented here that you want the Github Action to happen in this repo, so I have added the Dockerfile and a workflow yml file to this repo. I gave the workflow my best shot since I'm pretty unfamiliar with it - if you have suggestions to improve it, that would be great! I tested it with a Poplin Github repo pushing to a private Docker repo and it worked.

agnes-kiss commented:
Thanks! Giving it some thought it actually makes sense to add that as well to this repo so that we have everything in one place already. Not much experience with docker actually but have asked others to take a quick look as well and it looks good! (we can always improve with upcoming releases, if needed).
